Ensure Deployment labels adopted ReplicaSets and pods

janetkuo opened the pull request:
Ensure Deployment labels adopted ReplicaSets and pods. Otherwise, new RCs produced by Deployment will overlap we pre-existing ones that aren't constrained by the pod-template-hash.

Fixes #20755

Ignore the first commit (from #20994)
cc @bgrant0607 @nikhiljindal @ironcladlou @kargakis @kubernetes/sig-config @madhusudancs @mqliang

janetkuo committed:
Ensure Deployment labels adopted ReplicaSets and pods

bgrant0607 commented on pkg/util/labels/labels.go:
```diff
@@ -93,3 +106,30 @@ func CloneSelectorAndAddLabel(selector *unversioned.LabelSelector, labelKey stri
 }
+
+// AddLabelToSelector returns a selector with the given key and value added to the given selector's MatchLabels.
+func AddLabelToSelector(selector *unversioned.LabelSelector, labelKey string, labelValue uint32) *unversioned.LabelSelector {
```
In general, label values are strings, so this function should probably take a string labelValue.

janetkuo replied:
Fixed

bgrant0607 commented on pkg/util/labels/labels.go:
```diff
@@ -54,6 +54,19 @@ func CloneAndRemoveLabel(labels map[string]string, labelKey string) map[string]s
 }
 
+// AddLabel returns a map with the given key and value added to the given map.
+func AddLabel(labels map[string]string, labelKey string, labelValue uint32) map[string]string {
```
In general, label values are strings, so this function should probably take a string labelValue.

janetkuo replied:
Fixed

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +144,136 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+	}
+	syncedRSList := []extensions.ReplicaSet{}
+	for _, rs := range rsList {
+		// Add pod-template-hash information if it's not in the rs
```
This deserves more of an explanation about why it's important.

janetkuo replied:
Added explanation

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +144,136 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+			}
+			syncedRSList = append(syncedRSList, *updatedRS)
+		}
+		syncedRSList = append(syncedRSList, rs)
```
Is this adding the same RS twice?

kargakis replied:
@janetkuo missed a continue in the end of the if above but since that check is redundant, we will append only once.

janetkuo replied:
Yes, fixed now

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +144,136 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+	syncedRSList := []extensions.ReplicaSet{}
+	for _, rs := range rsList {
+		// Add pod-template-hash information if it's not in the rs
+		if !labelsutil.SelectorHasLabel(rs.Spec.Selector, extensions.DefaultDeploymentUniqueLabelKey) {
```
This is checked inside addHashKeyToReplicaSet. Does it need to be here?

kargakis replied:
@bgrant0607 is right, no need to check this twice

janetkuo replied:
Fixed

kargakis commented on pkg/util/deployment/deployment.go:
```diff
@@ -53,9 +55,12 @@ func GetOldReplicaSets(deployment extensions.Deployment, c clientset.Interface)
 		})
 }
 
+type rsListFunc func(string, api.ListOptions) ([]extensions.ReplicaSet, error)
```
Note that we have also been special-casing interfaces in our deployment controllers and now we are looking into switching back to full namespacers since they are much cleaner and more predictable imho. Why can't we use namespacers here too?

janetkuo replied:
I put these special-casing interfaces because `getOldReplciaSets` and `getNewReplicaSet` are called both in deployment controller and in deployment util. In controller we use rsStore and podStore to `List`, but in deployment util we use clientset.Interface to `List`. They're of different types (`storeXxxNamespacer` vs. `XxxInterface`) so I'm not sure how to avoid special-casing interfaces.

kargakis replied:
Nevermind. Add a TODO here about investigating to switching to full namespacers and I can look into it if I ever find time.

kargakis replied:
Actually can't we just use the client?

janetkuo replied:
Isn't using `List` of rsStore and podStore the whole point of having those cache stores?

kargakis replied:
Right. Please add a TODO for more investigation.

kargakis commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +144,136 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+		delay, maxRetries := 3, 3
+		for i := 0; i < maxRetries; i++ {
+			_, err = c.Core().Pods(namespace).Update(&pod)
+			if err != nil {
```
make this 
```go
if err == nil {
   break
}
```
to reduce nesting

janetkuo replied:
Done

kargakis commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +144,136 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+		return nil, err
+	}
+
+	// 4. Clean up any orphaned pods that don't have the new label, this can happen if the rs manager
```
Since we already account for orphaned pods, we could move pod relabeling at 1. and collapse updating the rs podtemplate and selector in one update?

janetkuo replied:
@kargakis nice suggestion! Done.

janetkuo committed:
Address comments; fix test failures; add e2e tests; update RS's label too

janetkuo commented:
It's ready for review now. PTAL @bgrant0607 @kargakis 

Addressed comments. Added e2e tests. I also add pod-template-hash to rs's labels (the first commit only added to rs's selector and template label).

kargakis commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +133,128 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		pod.Labels = labelsutil.AddLabel(pod.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
```
Should we check if the pod already has the label? It may happen that we start labeling some pods and then fail. I think we wouldn't want to relabel those pods.

janetkuo replied:
Fixed

kargakis commented:
@bgrant0607 I am worried that we are running the pod relabeling as part of an utility for deployments instead of in the rs controller loop. Can't we look up the pod-template-hash label in the rs controller and trigger the relabeling there?

mqliang commented on pkg/util/deployment/deployment.go:
```diff
@@ -25,9 +25,11 @@ import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/apis/extensions"
 	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	extensions_unversioned "k8s.io/kubernetes/pkg/client/typed/generated/extensions/unversioned"
```
IIRC, we should not use underline in imported package names. see https://github.com/kubernetes/kubernetes/issues/15319

janetkuo replied:
@mqliang noted. cc @caesarxuchao

caesarxuchao replied:
Thanks. I'll revise the other occurrences.

caesarxuchao replied:
Fixed in #21325

janetkuo replied:
Fixed

mqliang commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +133,128 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+				break
+			}
+			time.Sleep(time.Second * time.Duration(delay))
+			delay *= delay
```
I think we should add
```
pod, err = c.Core().Pods(namespace).Get(pod.Name)
``
here, otherwise Update will failed again.

janetkuo replied:
Fixed

bgrant0607 commented:
@kargakis Regarding doing the pod relabeling in the RS controller:

My immediate motivation for relabeling is so that adopted ReplicaSets don't overlap with ones created by a Deployment due to lack of the unique label. `kubectl rolling-update --image` did something similar.

An alternative would be to implement `controllerRef` in Deployment and ReplicaSet, which is described in #2210 and #14961. In that case, synchronizing `controllerRef` backpointers of the pods would be delegated to ReplicaSet.

That said, I think we'll want to propagate pod template label updates in the future. This is just one example of an in-place update: #9043. In-place updates make sense for cases where we want to minimize disruption. 

Even label updates can disrupt containers, in the case that they are visible to the containers via the downward API. In the future we may need to roll out such changes progressively, as we do for the regular rollout process.

I don't want ReplicaSet to need progressive rollout logic, as well. That logic should live in Deployment. 

If we think this is too tricky/risky, we could wait until the next release. With pause, adoption shouldn't be needed as often. We would need to document the problem this was trying to solve as a known issue with the beta implementation.

bgrant0607 commented:
I just wrote up a `controllerRef` proposal on #14961.

janetkuo committed:
Address comments

bgrant0607 commented:
Discussed in person: We're not going to try to implement `controllerRef` in 1.2. We'll go forward with this PR.

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+	// Copy the old selector, so that we can scrub out any orphaned pods
+	oldSelector := rs.Spec.Selector
+	// Update the selector of the rs so it manages all the pods we updated above
+	updatedRS, err := updateRSWithRetries(c.Extensions().ReplicaSets(namespace), &rs, func(updated *extensions.ReplicaSet) {
```
See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/rolling_updater.go#L642

That adds the new label to the pod template first in order to avoid races with the controller. A new pod without the label could be added between the loop above and this update.

janetkuo replied:
I combined updating pod template here because we can deal with orphaned pods in step 3. Changed it back to label pod template first.

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		// If the pod already has the new hash label, avoid re-labeling it
```
The label value could be different, in which case when we update the selector, it would orphan those pods. We might as well check the label value here as well.

janetkuo replied:
Fixed

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+			continue
+		}
+		pod.Labels = labelsutil.AddLabel(pod.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+		delay, maxRetries := 3, 3
```
Why only 3 retries? When retrying, we should distinguish between different failure reasons. For instance, conflicts and timeouts should be handled differently than more permanent conditions, like auth failure.

What would happen if this returned in the middle? 

In general, I'm concerned about the robustness and reentrancy of all the Deployment code. That's something we should test in the future. Please file an issue.

janetkuo replied:
Filed #21488; changed this to a retry function

If returned in the middle, this should be retried in the next sync, by checking if the rs selector contains the hash label or not (rs selector is updated the last so this check is solid). Plus, we check if the label is there before labeling things.

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+
+	// 3. Clean up any orphaned pods that don't have the new label, this can happen if the rs manager
+	//    doesn't see the update to its pod template and creates a new pod with the old labels after
+	//    we've finished re-adopting existing pods to the rs.
```
It would be better to add the label and let the controller decide what to do.

janetkuo replied:
Fixed

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+func updateRSWithRetries(rsClient unversionedextensions.ReplicaSetInterface, rs *extensions.ReplicaSet, applyUpdate updateFunc) (*extensions.ReplicaSet, error) {
+	var err error
+	oldRs := rs
+	err = wait.Poll(10*time.Millisecond, 1*time.Minute, func() (bool, error) {
```
Every client needs to do this. We should move this to the client library. I'll file an issue.

bgrant0607 replied:
Filed #21479

caesarxuchao replied:
Can we utilize the Unconditional Update here? I verified that the the strategy of ReplicaSet allows that.

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+		pod.Labels = labelsutil.AddLabel(pod.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+		delay, maxRetries := 3, 3
+		podName := pod.Name
+		for i := 0; i < maxRetries; i++ {
```
Also, this retry loop could be a helper function, similar to the RS one.

janetkuo replied:
Done

janetkuo committed:
Address comments

kargakis commented on pkg/util/deployment/deployment.go:
```diff
@@ -144,6 +136,158 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+		syncedRSList = append(syncedRSList, *syncedRS)
+	}
+	syncedPodList, err := getPodList(namespace, options)
+	return syncedRSList, syncedPodList, nil
```
return the error from above

janetkuo replied:
Fixed

kargakis commented on pkg/util/deployment/deployment.go:
```diff
@@ -144,6 +136,158 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+}
+
+// addHashKeyToRSAndPods adds pod-template-hash information to the given rs, if it's not already there, with the following steps:
+// 1. Add hash label to the rs's pod template
```
Is there a reason you split the steps back to 4? I would expect:

1. update all existing pods
2. update rs podtemplate and selector in one update
3. look for orphaned pods

to be the same plus being cheaper.

janetkuo replied:
Decided to change this to:
1. Update rs's pod template
1.5. Use observedGeneration to make sure the updated rs is seen by controller, so that no orphaned pods can be created after this step
2. Then, update pods
3. Last, update rs label and selector

This way, we avoid the clean up steps, and we can check rs's selector to know if we need to retry this labeling process in case we returned error and quit early. Since rs's selector is updated last, this check will be solid.

kargakis replied:
Cool

janetkuo committed:
Address comments; fix incorrect hash

janetkuo commented:
@bgrant0607 @kargakis PTAL

bgrant0607 commented:
```
# k8s.io/kubernetes/pkg/util/deployment
_output/local/go/src/k8s.io/kubernetes/pkg/util/deployment/deployment_test.go:30:2: cannot find package "k8s.io/kubernetes/pkg/client/testing/fake" in any of:
	/usr/local/go/src/k8s.io/kubernetes/pkg/client/testing/fake (from $GOROOT)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/testing/fake (from $GOPATH)
	/workspace/kubernetes/Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/testing/fake
FAIL	k8s.io/kubernetes/pkg/util/deployment [setup failed]
```

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -144,6 +137,166 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+		}
+	}
+	// Make sure rs pod template is updated so that it won't create pods without the new label (orphaned pods).
+	if updatedRS.Generation != updatedRS.Status.ObservedGeneration {
```
Should be >.

janetkuo replied:
Fixed

bgrant0607 commented on test/e2e/deployment.go:
```diff
@@ -797,3 +811,71 @@ func testRollbackDeploymentRSNoRevision(f *Framework) {
+		Expect(c.Extensions().ReplicaSets(ns).Delete(newRS.Name, nil)).NotTo(HaveOccurred())
+	}()
+
+	err = waitForDeploymentStatus(c, ns, deploymentName, replicas, replicas-1, replicas+1, 0)
```
I think this should work because the RS and pods should be relabeled before the status is updated by syncRollingUpdateDeployment. That's worth adding a comment about here and/or below, where the selector and labels of the RS are checked.

If that were not the case, we'd need a loop below to poll until the labels and selector were confirmed to be set.

janetkuo replied:
Added comment here

bgrant0607 commented on pkg/util/deployment/deployment_test.go:
```diff
@@ -26,10 +26,43 @@ import (
 	"k8s.io/kubernetes/pkg/apis/extensions"
 	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/fake"
+	"k8s.io/kubernetes/pkg/client/testing/core"
+	"k8s.io/kubernetes/pkg/client/testing/fake"
```
This looks like it moved into pkg/client/testing/core.

janetkuo replied:
Removed

bgrant0607 commented:
A couple minor comments and a test failure, but otherwise looks pretty good. Thanks.

kargakis commented on pkg/util/deployment/deployment.go:
```diff
@@ -144,6 +137,166 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+	return updatedRS, nil
+}
+
+func waitForReplicaSetUpdated(c clientset.Interface, desiredGeneration int64, namespace, name string) error {
```
Move this to pkg/client/conditions.go

janetkuo replied:
Do you mean `pkg/client/unversioned/conditions.go`? We use the generated client  instead of the unversioned one in this function, so this function should not be there. cc @caesarxuchao

kargakis replied:
Yes, sorry. How about creating `pkg/client/clientset_generated/internalclientset/conditions.go` then?

caesarxuchao replied:
No, because code in `pkg/client/clientset_generated` should be auto-generated. How about making a standalone `condition` package at `pkg/client/conditions/conditions.go`?

bgrant0607 replied:
Let's figure out the best place in another PR.

kargakis replied:
ok

janetkuo committed:
Address comments; fix test failure

janetkuo commented:
@bgrant0607 @kargakis Addressed comments and fixed the unit test failure; PTAL

bgrant0607 commented:
LGTM
